File: src/nse_calendar.cpp
#include "derivkit/nse_calendar.hpp"

#include <algorithm>
#include <array>
#include <cmath>
#include <iomanip>
#include <sstream>
#include <stdexcept>
#include <string>

namespace derivkit::nse {
namespace {

using namespace std::chrono;

constexpr int ymd_key(year_month_day ymd) {
    return static_cast<int>(ymd.year()) * 10000 +
           static_cast<int>(static_cast<unsigned>(ymd.month())) * 100 +
           static_cast<int>(static_cast<unsigned>(ymd.day()));
}

year_month_day ist_date(sys_seconds utc) {
    return year_month_day{floor<days>(utc + kIstOffset)};
}

minutes ist_tod(sys_seconds utc) {
    const auto ist = utc + kIstOffset;
    return duration_cast<minutes>(ist - floor<days>(ist));
}

sys_seconds ist_on(year_month_day ymd, minutes tod) {
    return sys_seconds{sys_days{ymd} + (tod - kIstOffset)};
}

bool is_weekend(year_month_day ymd) {
    const weekday wd{sys_days{ymd}};
    return wd == Saturday || wd == Sunday;
}

// NSE cash + F&O weekday holidays, calendar 2026 (Muhurat 8 Nov counted as open).
constexpr std::array<int, 16> kHolidays2026 = {
    20260115, 20260126, 20260303, 20260326, 20260331, 20260403, 20260414, 20260501,
    20260528, 20260626, 20260914, 20261002, 20261020, 20261110, 20261124, 20261225};

year_month_day prev_calendar_day(year_month_day ymd) {
    return year_month_day{sys_days{ymd} - days{1}};
}

}  // namespace

bool is_trading_day(year_month_day ymd) {
    if (!ymd.ok()) {
        return false;
    }
    const int key = ymd_key(ymd);
    if (key == 20261108) {  // Diwali Laxmi Pujan Muhurat session
        return true;
    }
    if (is_weekend(ymd)) {
        return false;
    }
    return std::find(kHolidays2026.begin(), kHolidays2026.end(), key) == kHolidays2026.end();
}

bool is_trading_day(const std::string& iso_date) { return is_trading_day(parse_iso_date(iso_date)); }

year_month_day parse_iso_date(const std::string& iso_date) {
    if (iso_date.size() < 10) {
        throw std::invalid_argument("date must be YYYY-MM-DD");
    }
    const int y = std::stoi(iso_date.substr(0, 4));
    const unsigned m = static_cast<unsigned>(std::stoi(iso_date.substr(5, 2)));
    const unsigned d = static_cast<unsigned>(std::stoi(iso_date.substr(8, 2)));
    const year_month_day ymd{year{y} / month{m} / day{d}};
    if (!ymd.ok()) {
        throw std::invalid_argument("invalid calendar date: " + iso_date);
    }
    return ymd;
}

sys_seconds expiry_close(const std::string& expiry_iso_date) {
    return ist_on(parse_iso_date(expiry_iso_date), kCloseIst);
}

sys_seconds snap_as_of(system_clock::time_point now) {
    auto utc = time_point_cast<seconds>(now);
    auto ymd = ist_date(utc);
    const minutes tod = ist_tod(utc);

    if (is_trading_day(ymd)) {
        if (tod >= kCloseIst) {
            return ist_on(ymd, kCloseIst);
        }
        if (tod >= kOpenIst) {
            return utc;
        }
    }
    ymd = prev_calendar_day(ymd);
    while (!is_trading_day(ymd)) {
        ymd = prev_calendar_day(ymd);
    }
    return ist_on(ymd, kCloseIst);
}

std::string format_ist(sys_seconds utc) {
    const auto ist = utc + kIstOffset;
    const year_month_day ymd{floor<days>(ist)};
    const auto tod = ist - sys_seconds{sys_days{ymd}};
    const auto hh = duration_cast<hours>(tod);
    const auto mm = duration_cast<minutes>(tod - hh);
    std::ostringstream oss;
    oss << static_cast<int>(ymd.year()) << '-' << std::setw(2) << std::setfill('0')
        << static_cast<unsigned>(ymd.month()) << '-' << std::setw(2)
        << static_cast<unsigned>(ymd.day()) << ' ' << std::setw(2) << hh.count() << ':'
        << std::setw(2) << mm.count() << " IST";
    return oss.str();
}

double trading_days_remaining(sys_seconds as_of, sys_seconds expiry) {
    if (as_of >= expiry) {
        return 1.0 / (kVolYearTradingDays * 24.0);  // one hour of vol time
    }
    double remaining = 0.0;
    const auto as_of_ymd = ist_date(as_of);
    if (is_trading_day(as_of_ymd)) {
        const auto close = ist_on(as_of_ymd, kCloseIst);
        if (as_of < close) {
            const auto open = ist_on(as_of_ymd, kOpenIst);
            const auto start = as_of < open ? open : as_of;
            const double hours = duration<double, std::ratio<3600>>(close - start).count();
            remaining += std::clamp(hours / kSessionHours, 0.0, 1.0);
        }
    }

    auto d = as_of_ymd;
    const auto expiry_ymd = ist_date(expiry);
    d = year_month_day{sys_days{d} + std::chrono::days{1}};
    while (d <= expiry_ymd) {
        if (is_trading_day(d)) {
            remaining += 1.0;
        }
        d = year_month_day{sys_days{d} + std::chrono::days{1}};
    }
    return std::max(remaining, 1.0 / (kVolYearTradingDays * 24.0));
}

Clock clock_to_expiry(const std::string& expiry_iso_date, system_clock::time_point as_of_in) {
    Clock c;
    const auto raw = time_point_cast<seconds>(as_of_in);
    c.as_of = snap_as_of(as_of_in);
    c.snapped_to_close = c.as_of != raw;
    c.expiry = expiry_close(expiry_iso_date);
    c.as_of_ist = format_ist(c.as_of);
    c.expiry_ist = format_ist(c.expiry);
    const double secs = duration<double>(c.expiry - c.as_of).count();
    c.time_rate = std::max(secs / (kRateYearDays * 24.0 * 3600.0), 1.0 / (kRateYearDays * 24.0));
    c.trading_days = trading_days_remaining(c.as_of, c.expiry);
    c.time_vol = c.trading_days / kVolYearTradingDays;
    return c;
}

}  // namespace derivkit::nse

File: src/derivkit/nse_calendar.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import date, datetime, timedelta, timezone
from typing import Union

IST = timezone(timedelta(hours=5, minutes=30))
UTC = timezone.utc

OPEN_IST = timedelta(hours=9, minutes=15)
CLOSE_IST = timedelta(hours=15, minutes=30)
SESSION_HOURS = 6.25
VOL_YEAR_TRADING_DAYS = 252.0
RATE_YEAR_DAYS = 365.25

_HOLIDAYS_2026 = {
    date(2026, 1, 15),
    date(2026, 1, 26),
    date(2026, 3, 3),
    date(2026, 3, 26),
    date(2026, 3, 31),
    date(2026, 4, 3),
    date(2026, 4, 14),
    date(2026, 5, 1),
    date(2026, 5, 28),
    date(2026, 6, 26),
    date(2026, 9, 14),
    date(2026, 10, 2),
    date(2026, 10, 20),
    date(2026, 11, 10),
    date(2026, 11, 24),
    date(2026, 12, 25),
}
_MUHURAT = date(2026, 11, 8)


def parse_iso_date(iso_date: str) -> date:
    if len(iso_date) < 10:
        raise ValueError("date must be YYYY-MM-DD")
    y = int(iso_date[0:4])
    m = int(iso_date[5:7])
    d = int(iso_date[8:10])
    try:
        return date(y, m, d)
    except ValueError as exc:
        raise ValueError(f"invalid calendar date: {iso_date}") from exc


def is_trading_day(ymd: Union[date, str]) -> bool:
    if isinstance(ymd, str):
        ymd = parse_iso_date(ymd)
    if ymd == _MUHURAT:
        return True
    if ymd.weekday() >= 5:
        return False
    return ymd not in _HOLIDAYS_2026


def _as_utc(ts: datetime) -> datetime:
    if ts.tzinfo is None:
        return ts.replace(tzinfo=UTC)
    return ts.astimezone(UTC)


def _ist_date(utc: datetime) -> date:
    return _as_utc(utc).astimezone(IST).date()


def _ist_tod(utc: datetime) -> timedelta:
    ist = _as_utc(utc).astimezone(IST)
    return timedelta(hours=ist.hour, minutes=ist.minute, seconds=ist.second, microseconds=ist.microsecond)


def _ist_on(ymd: date, tod: timedelta) -> datetime:
    local = datetime(ymd.year, ymd.month, ymd.day, tzinfo=IST) + tod
    return local.astimezone(UTC).replace(microsecond=0)


def expiry_close(expiry_iso_date: str) -> datetime:
    return _ist_on(parse_iso_date(expiry_iso_date), CLOSE_IST)


def snap_as_of(now: datetime) -> datetime:
    utc = _as_utc(now).replace(microsecond=0)
    ymd = _ist_date(utc)
    tod = _ist_tod(utc)
    if is_trading_day(ymd):
        if tod >= CLOSE_IST:
            return _ist_on(ymd, CLOSE_IST)
        if tod >= OPEN_IST:
            return utc
    ymd = ymd - timedelta(days=1)
    while not is_trading_day(ymd):
        ymd = ymd - timedelta(days=1)
    return _ist_on(ymd, CLOSE_IST)


def format_ist(utc: datetime) -> str:
    ist = _as_utc(utc).astimezone(IST)
    return f"{ist:%Y-%m-%d %H:%M} IST"


def _trading_days_remaining(as_of: datetime, expiry: datetime) -> float:
    if as_of >= expiry:
        return 1.0 / (VOL_YEAR_TRADING_DAYS * 24.0)
    remaining = 0.0
    as_of_ymd = _ist_date(as_of)
    if is_trading_day(as_of_ymd):
        close = _ist_on(as_of_ymd, CLOSE_IST)
        if as_of < close:
            open_t = _ist_on(as_of_ymd, OPEN_IST)
            start = open_t if as_of < open_t else as_of
            hours = (close - start).total_seconds() / 3600.0
            remaining += min(max(hours / SESSION_HOURS, 0.0), 1.0)
    d = as_of_ymd + timedelta(days=1)
    expiry_ymd = _ist_date(expiry)
    while d <= expiry_ymd:
        if is_trading_day(d):
            remaining += 1.0
        d += timedelta(days=1)
    return max(remaining, 1.0 / (VOL_YEAR_TRADING_DAYS * 24.0))


@dataclass
class Clock:
    as_of: datetime | None = None
    expiry: datetime | None = None
    as_of_ist: str = ""
    expiry_ist: str = ""
    time_rate: float = 0.0
    time_vol: float = 0.0
    trading_days: float = 0.0
    snapped_to_close: bool = False


def clock_to_expiry(expiry_iso_date: str, as_of: datetime | None = None) -> Clock:
    if as_of is None:
        as_of = datetime.now(tz=UTC)
    raw = _as_utc(as_of).replace(microsecond=0)
    c = Clock()
    c.as_of = snap_as_of(as_of)
    c.snapped_to_close = c.as_of != raw
    c.expiry = expiry_close(expiry_iso_date)
    c.as_of_ist = format_ist(c.as_of)
    c.expiry_ist = format_ist(c.expiry)
    secs = (c.expiry - c.as_of).total_seconds()
    c.time_rate = max(secs / (RATE_YEAR_DAYS * 24.0 * 3600.0), 1.0 / (RATE_YEAR_DAYS * 24.0))
    c.trading_days = _trading_days_remaining(c.as_of, c.expiry)
    c.time_vol = c.trading_days / VOL_YEAR_TRADING_DAYS
    return c
